### DISCUSSION TITLE
#include documentation

### DISCUSSION THREAD

**Discussion by geoffthemedio** [wjakob/nanobind#143] [Q&A] [answered]
In [basics.rst](https://github.com/wjakob/nanobind/blob/master/docs/basics.rst) file, it is noted, "_nanobind also provides many optional add-on components that are aren't included by default. They are discussed throughout this documentation along with pointers to the header files that must be included when using them._" Later in the file, in the "Higher order functions" section, `nb::print` is used, but it doesn't indicate what header to `#include` to declare this function. If viewed at [nanobind.readthedocs.io](https://nanobind.readthedocs.io/en/latest/basics.html), there are links to the [API documentation](https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4N8nanobind5printE6handle6handle6handle), but that also doesn't indicate what `#include` is needed for each function.

I searched through the headers, and found [inline defintions for `print` in nb_types.h](https://github.com/wjakob/nanobind/blob/master/include/nanobind/nb_types.h#L507). But when I try including that in my teste extension .cpp file, I get many build errors and messages, eg.

````
nb_types.h(64,10): warning C4348: 'nanobind::borrow': redefinition of default parameter: parameter 1
nb_types.h(64,34): message : see declaration of 'nanobind::borrow'
nb_types.h(65,10): warning C4348: 'nanobind::steal': redefinition of default parameter: parameter 1
nb_types.h(65,34): message : see declaration of 'nanobind::steal'
nb_types.h(69,10): warning C4348: 'nanobind::detail::type_caster': redefinition of default parameter: parameter 2
nb_types.h(69,53): message : see declaration of 'nanobind::detail::type_caster'
nb_types.h(77,17): error C2011: 'nanobind::detail::borrow_t': 'struct' type redefinition
nb_types.h(77,8): message : see declaration of 'nanobind::detail::borrow_t'
nb_types.h(78,16): error C2011: 'nanobind::detail::steal_t': 'struct' type redefinition
````

I suspect I'm not including the right file, or not in the right order, but I can't find what I should be including. It would be helpful if the example in "basics" included the `#include`, and if the API documentation indicated this as well (other than the single #include indication that is in the api_core page for `#include <nanobind/operators.h>`). For example, cppreference has "Defined in header" at the top of the pages for various standard library features, which is very helpful.

**Comment by wjakob** [comment, marked as answer]
The `nb_*` headers aren't meant to be included manually -- they are all automatically pulled in by `nanobind.h` (and in the right order). These headers are also never mentioned in the documentation.  You should be able to use `nb::print` after including `nanobind.h`
